packages/grid/_modules_/grid/hooks/features/core/useGridState.ts

```diff
@@ -18,29 +18,25 @@ export const useGridState = (
   const setGridState = React.useCallback(
     (stateUpdaterFn: (oldState: GridState) => GridState) => {
       const newState = stateUpdaterFn(apiRef.current.state);
-      const hasChanged = apiRef.current.state !== newState;
-      if (!hasChanged) {
-        // We always assign it as we mutate rows for perf reason.
-        apiRef.current.state = newState;
+      if (apiRef.current.state === newState) {
         return false;
       }
 
-      const { shouldUpdate, postUpdate } = apiRef.current.applyControlStateConstraint(newState);
-
-      if (!shouldUpdate) {
-        return false;
-      }
-      // We always assign it as we mutate rows for perf reason.
-      apiRef.current.state = newState;
+      const { ignoreSetState, postUpdate } = apiRef.current.applyControlStateConstraint(newState);
 
-      if (hasChanged && apiRef.current.publishEvent) {
-        const params: GridStateChangeParams = { api: apiRef.current, state: newState };
-        apiRef.current.publishEvent(GRID_STATE_CHANGE, params);
+      if (!ignoreSetState) {
+        // We always assign it as we mutate rows for perf reason.
+        apiRef.current.state = newState;
 
-        postUpdate();
+        if (apiRef.current.publishEvent) {
+          const params: GridStateChangeParams = { api: apiRef.current, state: newState };
+          apiRef.current.publishEvent(GRID_STATE_CHANGE, params);
+        }
       }
 
-      return hasChanged;
+      postUpdate();
```

Comment:
It should only be called if `ignoreSetState` is false. The following test passes on HEAD but here it doesn't.

```diff
diff --git a/packages/grid/x-grid/src/tests/selection.XGrid.test.tsx b/packages/grid/x-grid/src/tests/selection.XGrid.test.tsx
index f709fad2..4c453247 100644
--- a/packages/grid/x-grid/src/tests/selection.XGrid.test.tsx
+++ b/packages/grid/x-grid/src/tests/selection.XGrid.test.tsx
@@ -15,6 +15,7 @@ import {
   GridSelectionModel,
   useGridApiRef,
   XGrid,
+  GRID_SELECTION_CHANGE
 } from '@material-ui/x-grid';

 const isJSDOM = /jsdom/.test(window.navigator.userAgent);
@@ -322,5 +323,15 @@ describe('<XGrid /> - Selection', () => {
       expect(getRow(1)).to.have.class('Mui-selected');
       expect(getRow(2)).to.have.class('Mui-selected');
     });
+
+    it('should not publish GRID_SELECTION_CHANGE if the selection state did not change ', () => {
+      const onSelectionChange = spy();
+      const selectionModel = [];
+      render(<Test selectionModel={selectionModel} />);
+      apiRef.current.subscribeEvent(GRID_SELECTION_CHANGE, onSelectionChange);
+      expect(onSelectionChange.callCount).to.equal(0);
+      apiRef.current.setSelectionModel(selectionModel);
+      expect(onSelectionChange.callCount).to.equal(0);
+    });
   });
 });
```

Reply:
Ah, well spotted, I'm adding a new commit to fix this.